### PR TITLE
Fix linebreaks

### PR DESCRIPTION
**Pull request by ixs**
Soft Line Breaks are not handled optimally in Mail_Mime.

* Breaks happen too early at char 76 rather than _after_ char 76 meaning that a line gets broken up too early.
* In the situation that a dot would end up as the first char on a new line extra handling and extra escaping is needed. This calculation suffers from another miscalculation reducing the line length by one more char.

Fix both these issues.
Add tests for to both issues.

### CODE REVIEW

**Comment by ixs** [comment]
Mhm. Just found another weird edge case that fails. Need to have another look and figure it out...

**Comment by ixs** [comment]
Alright. Better now. 😁

File: tests/encoding_case.phpt [hunk 1]
```diff
@@ -8,6 +8,5 @@ require_once('Mail/mimePart.php');
 print Mail_mimePart::quotedPrintableEncode($test, 7);
 ?>
 --EXPECT--
-Here's=
-=09
+Here's=09
```
Comment (till) [inline tests/encoding_case.phpt:11]: @ixs Email is not really my problem domain anymore. I see you adjusted another test, is that save to do?

@pear/mail-components Anyone else wants to weigh in? LGTM otherwise.

File: tests/test_linebreak_dot.phpt [hunk 2]
```diff
@@ -0,0 +1,48 @@
+--TEST--
```
Comment (till) [inline tests/test_linebreak_dot.phpt:1]: 🎉 Yay, new test!

File: tests/test_linebreak_larger_76.phpt [hunk 3]
```diff
@@ -0,0 +1,59 @@
+--TEST--
```
Comment (till) [inline tests/test_linebreak_larger_76.phpt:1]: 👍

**Review by till — APPROVED** [review]

File: tests/encoding_case.phpt (hunk 1 above)
Comment (ixs) [reply tests/encoding_case.phpt:11]: @till 😁 That test broke after my change.
I had a hard and long look at it, fixed a bug in my code and then looked again at the test.

I was just explaining what exactly the old code did and why it did it wrong here... But while writing that down, I'm having a few doubts about and think I might have missed an interesting case... So let me have another hard look at this test and get back to you in a few. 😝

**Comment by ixs** [comment]
Pushed a new version.

@till So I had another hard look at the code and I missed one case indeed. I did not read the tab test results correctly initially. So thanks for asking and making me have another look. 👍🏻

Reverted the change to the tests/encoding_case.phpt file and added one more test file that I used during debugging. It's partly duplicating existing tests but it contains a nice print function that is good for human consumption...

**Review by till — APPROVED** [review]
This LGTM.

Maybe @alecpl wants to take a look?

File: tests/qp_encoding_test.phpt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        $input = substr($text, 0, $i);
+        $output = Mail_mimePart::quotedPrintableEncode($input, $break);
+
+        printf("input: %02d: %s\n", strlen($input), $input);
+
+        $lines = explode("\r\n", $output);
+        for($j=0; $j < count($lines); $j++) {
+            $line = $lines[$j];
+            if ($j + 1 < count($lines) && $special_chars) {
+                $line_vis = str_replace("\t", '\t', $line).'\r\n';
```
Comment (alecpl) [inline tests/qp_encoding_test.phpt:35]: Would be nice to have this str_replace() used for printing the input too.

File: tests/qp_encoding_test.phpt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+output:10: 123456789=
+output:04: =2E1
+---
+input: 12: 123456789.12
+output:10: 123456789=
+output:05: =2E12
+---
+input: 15: 	Here's	
+a tab.
+output:08: 	Here's=
```
Comment (alecpl) [inline tests/qp_encoding_test.phpt:95]: I would expect to see `\t` here in the start of output.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: tests/qp_encoding_test.phpt (hunk 4 above)
Comment (ixs) [reply tests/qp_encoding_test.phpt:40]: It makes it easier to read when tabs are involved, but with newlines it's now one long string...

But ehh, it's six of one, half a dozen of the other... Added the change.
